test/adaptive_load/metrics_evaluator_test.cc

```diff
@@ -0,0 +1,368 @@
+  absl::StatusOr<BenchmarkResult> result_or = evaluator.AnalyzeNighthawkBenchmark(
+      bad_nighthawk_response, spec, name_to_custom_metrics_plugin_map);
+  ASSERT_FALSE(result_or.ok());
+  EXPECT_EQ(result_or.status().code(), absl::StatusCode::kUnavailable);
+  EXPECT_EQ(result_or.status().message(), kExpectedErrorMessage);
+}
+
+TEST(AnalyzeNighthawkBenchmark, StoresNighthawkResultForSuccessfulMetricEvaluation) {
+  nighthawk::adaptive_load::AdaptiveLoadSessionSpec spec;
+  nighthawk::client::ExecutionResponse nighthawk_response = MakeNighthawkResponseWithSendRate(1.0);
+  absl::flat_hash_map<std::string, MetricsPluginPtr> name_to_custom_metrics_plugin_map;
+
+  MetricsEvaluatorImpl evaluator;
+  absl::StatusOr<BenchmarkResult> result_or = evaluator.AnalyzeNighthawkBenchmark(
+      nighthawk_response, spec, name_to_custom_metrics_plugin_map);
+  ASSERT_TRUE(result_or.ok());
+
+  EXPECT_TRUE(MessageDifferencer::Equivalent(result_or.value().nighthawk_service_output(),
+                                             nighthawk_response.output()));
+  EXPECT_EQ(result_or.value().nighthawk_service_output().DebugString(),
+            nighthawk_response.output().DebugString());
+}
+
+TEST(AnalyzeNighthawkBenchmark, StoresScoreForSuccessfulMetricEvaluation) {
+  nighthawk::adaptive_load::AdaptiveLoadSessionSpec spec;
+
+  const std::string kMetricName = "good-metric";
+  const double kExpectedValue = 123.0;
+
+  FakeMetricsPluginConfig metrics_plugin_config;
+  FakeMetricsPluginConfig::FakeMetric* fake_metric =
+      metrics_plugin_config.mutable_fake_metrics()->Add();
+  fake_metric->set_name(kMetricName);
+  fake_metric->set_value(kExpectedValue);
+
+  MetricSpec metric_spec;
+  metric_spec.set_metrics_plugin_name("nighthawk.fake_metrics_plugin");
+  metric_spec.set_metric_name(kMetricName);
+
+  *spec.mutable_informational_metric_specs()->Add() = metric_spec;
+
+  nighthawk::client::ExecutionResponse nighthawk_response = MakeNighthawkResponseWithSendRate(1.0);
+  absl::flat_hash_map<std::string, MetricsPluginPtr> name_to_custom_metrics_plugin_map;
+  name_to_custom_metrics_plugin_map["nighthawk.fake_metrics_plugin"] =
+      std::make_unique<FakeMetricsPlugin>(metrics_plugin_config);
+
+  MetricsEvaluatorImpl evaluator;
+  absl::StatusOr<BenchmarkResult> result_or = evaluator.AnalyzeNighthawkBenchmark(
+      nighthawk_response, spec, name_to_custom_metrics_plugin_map);
+  ASSERT_TRUE(result_or.ok());
+  ASSERT_GT(result_or.value().metric_evaluations().size(), 0);
+  EXPECT_EQ(result_or.value().metric_evaluations()[0].metric_value(), kExpectedValue);
+}
+
+TEST(AnalyzeNighthawkBenchmark, PropagatesErrorFromFailedMetricEvaluation) {
+  nighthawk::adaptive_load::AdaptiveLoadSessionSpec spec;
+
+  const std::string kMetricName = "bad-metric";
+  const std::string kExpectedStatusMessage = "artificial metric error";
+
+  FakeMetricsPluginConfig metrics_plugin_config;
+  FakeMetricsPluginConfig::FakeMetric* fake_metric =
```

Comment:
Optional:  It seems like we're doing this same pattern a lot, and it takes up a lot of code (lack of textproto parsing).  Any way to create helper functions to reduce some of the lines of boilerplate?

Reply:
There are two occurrences here, with one having a value and one having an error, so we wouldn't save anything in this test by adding helpers that would only get called once each. The reason we don't have helpers in `FakeMetricsPlugin` is that it seemed like too much work after `FakeMetricsPlugin` became more flexible. It would be great if we could get textproto parsing working.  Opened #520.